# Practice using OpenStreetMap
## How I created gas_stations.csv

Create a working, clean dataframe of all gas stations in CA. Columns:
- Name
- Latitude
- Longitude

Columns to add:
- Heat Category

In [1]:
import requests
import pandas as pd

In [12]:
overpass_url = "http://overpass-api.de/api/interpreter"
overpass_query = """
    [out:json];
    area[name = "California"]->.search;
    node[amenity='fuel'](area.search);
    out count;
"""

response = requests.get(overpass_url, params={'data': overpass_query})
response.json()

{'version': 0.6,
 'generator': 'Overpass API 0.7.56.6 474850e8',
 'osm3s': {'timestamp_osm_base': '2020-10-27T19:06:03Z',
  'timestamp_areas_base': '2020-10-27T18:38:03Z',
  'copyright': 'The data included in this document is from www.openstreetmap.org. The data is made available under ODbL.'},
 'elements': [{'type': 'count',
   'id': 0,
   'tags': {'nodes': '2927',
    'ways': '0',
    'relations': '0',
    'areas': '0',
    'total': '2927'}}]}

### There are a total of 2,927 gas stations in CA

In [13]:
overpass_url = "http://overpass-api.de/api/interpreter"
overpass_query = """
    [out:json];
    area[name = "California"]->.search;
    node[amenity='fuel'](area.search);
    out;
"""
# specify a bounding box by (south, west, north, east) in latitude and longitude
response = requests.get(overpass_url, params={'data': overpass_query})
response.json()

{'version': 0.6,
 'generator': 'Overpass API 0.7.56.6 474850e8',
 'osm3s': {'timestamp_osm_base': '2020-10-27T19:11:02Z',
  'timestamp_areas_base': '2020-10-27T18:38:03Z',
  'copyright': 'The data included in this document is from www.openstreetmap.org. The data is made available under ODbL.'},
 'elements': [{'type': 'node',
   'id': 97044219,
   'lat': 34.0573651,
   'lon': -117.1907772,
   'tags': {'addr:housenumber': '765',
    'addr:street': 'W Redlands Blvd',
    'amenity': 'fuel',
    'description': 'Average fuel prices, best prices on fountain drinks on area. Friendly staff',
    'name': 'Circle K & Minimart',
    'opening_hours': '24/7',
    'operator': 'Circle K'}},
  {'type': 'node',
   'id': 98659556,
   'lat': 34.0485747,
   'lon': -117.2256087,
   'tags': {'amenity': 'fuel', 'created_by': 'Potlatch 0.5'}},
  {'type': 'node',
   'id': 140103093,
   'lat': 34.0413876,
   'lon': -117.1562993,
   'tags': {'amenity': 'fuel', 'created_by': 'Potlatch 0.5d'}},
  {'type': 'node',
 

In [20]:
len(response.json()['elements'])

2927

In [26]:
response.json()['elements'][10]['tags']['name']

'Car Service'

# Create columns:
- Name
- Latitude
- Longitude

In [29]:
names = []
for x in range(len(response.json()['elements'])):
    if 'name' in response.json()['elements'][x]['tags']:
        names.append(response.json()['elements'][x]['tags']['name'])
    else:
        names.append(None)
names

['Circle K & Minimart',
 None,
 None,
 'Arco',
 'Valley Oil',
 None,
 'Arco',
 'Arco',
 'Shell',
 '76',
 'Car Service',
 'Chevron',
 'Mobil',
 'Fuel 24:7',
 None,
 'Chevron',
 'Chevron',
 None,
 'Valero',
 'Chevron',
 None,
 None,
 None,
 'Valero',
 '76',
 'Shell',
 None,
 'Shell',
 'Chevron',
 'World',
 'Shell',
 'Mobil',
 None,
 'Clairemont Auto Care',
 None,
 None,
 'Chevron',
 'Chevron Mission Car Wash',
 'Arco',
 'Patriot',
 'Arco',
 'Arco',
 'Shell',
 'Shell',
 'Touchless Gasoline',
 'Arco',
 'Chevron',
 'Chevron',
 None,
 'Shell',
 None,
 'Chevron',
 None,
 None,
 None,
 None,
 None,
 'Costco Gasoline',
 None,
 'Arco',
 'Arco',
 '76',
 'Cal Gas',
 'Arco',
 None,
 'Chevron',
 'Shell',
 'Rotten Robbie #31',
 'Chevron',
 'Mobil',
 None,
 'Chevron',
 '7-Eleven',
 'Chevron',
 'Chevron',
 None,
 None,
 '76',
 'Valero',
 None,
 'Chevron',
 'Chevron',
 'Shell',
 'VP',
 "Jesse's",
 'Golden Eagle',
 'Quest Mart',
 'Shell',
 'Chevron',
 'Chevron',
 'Mobil',
 '76',
 'Arco',
 '76',
 '76 Gas 

In [31]:
names.count(None)

758

#### There are a lot of gas stations without names. bummer

In [15]:
latitudes = []
for x in range(len(response.json()['elements'])):
    latitudes.append(response.json()['elements'][x]['lat'])
latitudes

[34.0573651,
 34.0485747,
 34.0413876,
 37.7684682,
 37.381165,
 34.1032728,
 37.4060647,
 37.6774981,
 37.6777037,
 37.8373034,
 37.8407536,
 37.8404305,
 37.8394465,
 37.8400252,
 37.7755014,
 37.8255858,
 37.8579312,
 37.84382,
 37.581745,
 37.5803881,
 37.5872485,
 37.5876457,
 37.578398,
 37.5887203,
 37.2210474,
 33.7165414,
 37.8385166,
 37.3731217,
 32.8670766,
 32.8548259,
 32.8543944,
 32.8542675,
 32.8227832,
 32.8342381,
 32.8342381,
 32.8348112,
 38.5510626,
 34.0909625,
 37.6003825,
 40.0093833,
 37.551753,
 34.0635637,
 34.0655247,
 34.0646722,
 37.8684979,
 38.546829,
 34.0623716,
 37.4792248,
 39.2662359,
 37.764585,
 33.8798779,
 33.831782,
 33.7423831,
 33.7467724,
 37.3382722,
 33.887245,
 33.8890904,
 37.6405002,
 37.8554553,
 37.8503278,
 37.7621609,
 37.7665265,
 37.3268208,
 37.3411416,
 37.3540568,
 39.1913805,
 39.151165,
 37.0210099,
 33.7158403,
 33.7158375,
 37.3524507,
 37.0008405,
 37.3406041,
 37.3775792,
 36.984475,
 36.9848483,
 37.7798017,
 38.8352662

In [16]:
longitudes = []
for x in range(len(response.json()['elements'])):
    longitudes.append(response.json()['elements'][x]['lon'])
longitudes

[-117.1907772,
 -117.2256087,
 -117.1562993,
 -122.420268,
 -122.0701271,
 -118.2372374,
 -122.078426,
 -121.7879328,
 -121.7868756,
 -122.2503714,
 -122.2372221,
 -122.2615441,
 -122.2700151,
 -122.2701261,
 -122.2334503,
 -122.2097519,
 -122.2441212,
 -122.2467567,
 -122.35037,
 -122.3341399,
 -122.3634177,
 -122.3623912,
 -122.3371847,
 -122.3627774,
 -121.8623645,
 -117.954905,
 -120.230955,
 -122.087399,
 -117.2158525,
 -117.2041575,
 -117.204072,
 -117.2049467,
 -117.1843763,
 -117.1950836,
 -117.1944828,
 -117.1950936,
 -121.6964249,
 -118.0903715,
 -122.3831517,
 -123.7905532,
 -122.30565,
 -117.2255432,
 -117.2264817,
 -117.2092079,
 -122.2664016,
 -121.7593854,
 -117.1808061,
 -122.1518909,
 -123.2041124,
 -122.4243389,
 -118.1429756,
 -118.0875412,
 -118.099086,
 -118.105936,
 -122.009687,
 -118.105441,
 -118.1249503,
 -122.4015096,
 -122.2668563,
 -122.260754,
 -122.244179,
 -122.241209,
 -121.8752593,
 -121.879584,
 -122.0887173,
 -123.2084302,
 -123.1993253,
 -121.5694177

# Create dataframe

In [33]:
gas_df = pd.DataFrame(data = list(zip(names, longitudes, latitudes)), columns = ['name', 'longitude', 'latitude'])

In [34]:
gas_df.head()

name   longitude   latitude
0  Circle K & Minimart -117.190777  34.057365
1                 None -117.225609  34.048575
2                 None -117.156299  34.041388
3                 Arco -122.420268  37.768468
4           Valley Oil -122.070127  37.381165

In [36]:
gas_df.to_csv('./data/gas_stations.csv', index = False)